### DISCUSSION TITLE
Complete No-API Self-Host Option

### DISCUSSION THREAD

**Discussion by BradKML** [firecrawl/firecrawl#195] [Ideas]
Hello from another project! Can this repo be run with a local scraper and reliance on Ollama instead of paywalled systems? https://github.com/langgenius/dify/discussions/4565

**Comment by mattjoyce** [comment]
You can run the scraper locally, but the structured output is designed for openai from what I can see,
I also would like an Ollama option and while I think I could implement it, I'm not sure which model Ollama can run that would be good for OpenAI style json schemas.

I think the problem will often be context size, this project uses 32k.

**Reply by BradKML** [reply]
Yeah I was kinda concerned if the scraper is really SaaS or not, since for small tasks with full-blog crawling you don't need to get too fancy (maybe a few anti-bot dodge tactics but that is it)
Also for the context size, would using a RAG (embeddings and rerankers) or a large context model be able to help with the situation? How far is this from wget caching or ArchiveBox (similar to HTTRack)? https://github.com/ArchiveBox/ArchiveBox/issues/191
Small reference to browser fingerprint bypassing: https://github.com/niespodd/browser-fingerprinting

**Comment by appinteractive** [comment]
You could try https://ollama.com/library/nexusraven with function calling. There might be other models that have bigger contexts but I would need to research further. 

And ollana is fully openai compatible to my knowledge.

**Reply by BradKML** [reply]
Ollama is truly the go-to option but they still lagging over rerankers. For context windows Command-R and Gradient's finetuning are probs the best bet... Hopefully more models are supported in the future.

**Reply by appinteractive** [reply]
I think there are fine tuned versions of mistral with larger context windows. But I don't think they use 0.3 already with function calling.

**Comment by mattjoyce** [comment]
Mistral 0.3 was released recently, they have included structured output in that version.
However, It think with scraping context size is critical.

I'm going to have a go to include Ollama, and test.

**Reply by BradKML** [reply]
Could you document how Firecrawl can work with Python and Ollama?

**Reply by mattjoyce** [reply]
firecrawl has integrated llm calling in the api, it will be  fair amount of work to make it natively work with Ollama.
but if you want to use firecrawl *with* Ollama you can do something like this....

```
from ollama import Client
import os
from typing import Optional
from firecrawl import FirecrawlApp

os.environ["FIRECRAWL_API_URL"] = "http://localhost:3002"
os.environ["FIRECRAWL_API_KEY"] = "your_api_key"
app=FirecrawlApp()
url = "https://www.humblebundle.com/books/cicd-mastery-for-cloud-infrastructure-packt-books"
params = {
    'pageOptions': {
        'onlyMainContent': True
    }
}

scraped_data = app.scrape_url(url)
prompt1 = f"""TASK -> Summerize this scraped webpage
CONTENT -> {scraped_data['markdown']}
"""


client = Client(host="http://localhost:11434")
response = client.chat(
    model="mistral:latest",
    keep_alive=600,
    options={"temperature": 0},
    messages=[
        {
            "role": "system",
            "content": "You excel at reading and interpreting scraped webpage data",
            "role": "user",
            "content": prompt1,
        },
    ],
)
response1 = response["message"]["content"]
print(scraped_data)
print(response1)
```

but that is not structured output.

**Comment by FingerLiu** [comment]
just use new llama 3.1， I tested it it works fine.

**Reply by BradKML** [reply]
That is good to know, including function calls right?

**Reply by FingerLiu** [reply]
yep

**Reply by BradKML** [reply]
Damn this needs to be integrated into systems like OpenDevin then

**Comment by kinthaiofficial** [comment]
Self-hosting scrapers for agent pipelines makes sense — agents that can browse and extract web content without external API dependencies are more autonomous and have lower latency.

A few things that matter for agent-oriented web scraping:

**Rate limiting as first-class config** — agents will hammer scrapers if you let them. The self-hosted version needs configurable rate limits per agent (not just per IP) so a single runaway agent can't exhaust the crawl budget for the whole system.

**Extraction with schema** — raw HTML → markdown is useful, but for agents, "extract this page into `ProductSchema`" is what you actually want. Self-hosted Firecrawl with pluggable extraction schemas (JSON Schema or Pydantic) would let agents specify exactly what data they need rather than getting a full page dump and parsing it themselves.

**Provenance metadata** — when an agent extracts a fact from a scraped page, it should be able to cite the exact URL, timestamp, and HTML element path. This is the source citation chain that makes agent outputs verifiable.

**Async crawl jobs** — agents shouldn't block waiting for a scrape. Submit a crawl job, get a job ID, receive a webhook or poll for completion. This lets the agent proceed with other sub-tasks while the crawl runs in the background.

**Cost visibility** — agents need to know how much a crawl costs (in time and compute) before deciding to do it. A cost estimate endpoint would let the agent decide "is this crawl worth it given my remaining budget?"

We've integrated Firecrawl into KinthAI's agent network for web-aware agent tasks: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

Is the no-API self-host request primarily about privacy, cost, or reliability?
